core/src/iam/verify.rs

```diff
@@ -164,41 +165,72 @@ pub async fn token(kvs: &Datastore, session: &mut Session, token: &str) -> Resul
 			// Create a new readonly transaction
 			let mut tx = kvs.transaction(Read, Optimistic).await?;
 			// Parse the record id
-			let id = syn::thing(&id)?;
+			let mut rid = syn::thing(&id)?;
 			// Get the database access method
 			let de = tx.get_db_access(&ns, &db, &ac).await?;
 			// Obtain the configuration to verify the token based on the access method
-			let cf = match de.kind {
-				AccessType::Record(ac) => match ac.jwt.verify {
-					JwtAccessVerify::Key(key) => config(key.alg, key.key),
-					#[cfg(feature = "jwks")]
-					JwtAccessVerify::Jwks(jwks) => {
-						if let Some(kid) = token_data.header.kid {
-							jwks::config(kvs, &kid, &jwks.url, token_data.header.alg).await
-						} else {
-							Err(Error::MissingTokenHeader("kid".to_string()))
+			let (at, cf) = match de.kind {
+				AccessType::Record(at) => {
+					let cf = match at.jwt.verify.clone() {
+						JwtAccessVerify::Key(key) => config(key.alg, key.key),
+						#[cfg(feature = "jwks")]
+						JwtAccessVerify::Jwks(jwks) => {
+							if let Some(kid) = token_data.header.kid {
+								jwks::config(kvs, &kid, &jwks.url, token_data.header.alg).await
+							} else {
+								Err(Error::MissingTokenHeader("kid".to_string()))
+							}
 						}
-					}
-					#[cfg(not(feature = "jwks"))]
-					_ => return Err(Error::AccessMethodMismatch),
-				},
+						#[cfg(not(feature = "jwks"))]
+						_ => return Err(Error::AccessMethodMismatch),
+					}?;
+
+					(at, cf)
+				}
 				_ => return Err(Error::AccessMethodMismatch),
-			}?;
+			};
 			// Verify the token
 			decode::<Claims>(token, &cf.0, &cf.1)?;
+			// AUTHENTICATE clause
+			if let Some(ac) = at.authenticate {
+				// Setup the system session for finding the signin record
+				let mut sess = Session::editor().with_ns(&ns).with_db(&db);
+				sess.rd = Some(rid.clone().into());
+				sess.tk = Some(token_data.claims.clone().into());
+				sess.ip.clone_from(&session.ip);
+				sess.or.clone_from(&session.or);
+				// Compute the value with the params
+				match kvs.evaluate(ac, &sess, None).await {
+					Ok(val) => match val.record() {
+						Some(id) => {
+							// Update rid with result from AUTHENTICATE clause
+							rid = id;
+							// We don't generate the token in this scenario, so unlike signin/signup, no need to update claims.
```

Comment:
I think this exception makes my previous comment all the more relevant. We should be consistent on whether or not we keep the `tk` session attribute intact or not. I would much prefer to keep it intact.

Reply:
I addressed this in [a54b5bd](https://github.com/surrealdb/surrealdb/pull/4161/commits/a54b5bd5c4a8b4817a71b8f90a2cf8ba87ea0949). Let me know what you think!